Sources/SourceKitLSP/Swift/CodeActions/ConvertStringConcatenationToStringInterpolation.swift

```diff
@@ -0,0 +1,184 @@
+  /// expression for string interpolation. `exprList` must also contain at least one string literal child.
+  ///
+  ///  The following is a valid string concatenation.
+  ///``` swift
+  ///"Hello " + aString + "\(1)World"
+  ///```
+  ///  The following are invalid string concatenations.
+  ///``` swift
+  ///aString + bString // no string literals
+  ///
+  ///"Hello " * aString - "World" // non `+` operators
+  ///
+  ///"""
+  ///Hello
+  ///"""
+  ///+ """
+  ///World
+  ///""" // multi-line string literals
+  ///```
+  private static func preflight(
+    exprList: ExprListSyntax
+  ) -> (componentsOnly: [ExprListSyntax.Element], longestPounds: TokenSyntax?)? {
+    var iter = exprList.makeIterator()
+    guard let first = iter.next() else {
+      return nil
+    }
+
+    var hasStringLiterals = false
+    var longestPounds: TokenSyntax?
+    var componentsOnly = [first]
+
+    if let stringLiteral = first.as(StringLiteralExprSyntax.self) {
+      guard stringLiteral.isSingleLine else {
+        return nil
+      }
+      hasStringLiterals = true
+      longestPounds = stringLiteral.openingPounds
+    }
+
+    while let concat = iter.next(), let stringComponent = iter.next() {
+      guard let concat = concat.as(BinaryOperatorExprSyntax.self),
+        concat.isConcat && !stringComponent.is(MissingExprSyntax.self)
+      else {
+        return nil
+      }
+
+      if let stringLiteral = stringComponent.as(StringLiteralExprSyntax.self) {
+        guard stringLiteral.isSingleLine else {
+          return nil
+        }
+        hasStringLiterals = true
+        if let pounds = stringLiteral.openingPounds, pounds.isLongerThanAsTrimmed(longestPounds) {
+          longestPounds = pounds
+        }
+      }
+
+      componentsOnly.append(stringComponent)
+    }
+
+    guard hasStringLiterals && componentsOnly.count > 1 else {
+      return nil
+    }
+
+    return (componentsOnly, longestPounds)
+  }
+}
+
+extension ConvertStringConcatenationToStringInterpolation: SyntaxRefactoringCodeActionProvider {
+  static let title: String = "Convert String Concatenation to String Interpolation"
+
+  static func nodeToRefactor(in scope: SyntaxCodeActionScope) -> SequenceExprSyntax? {
+    guard let token = scope.innermostNodeContainingRange,
+      let seqExpr = token.findParentOfSelf(
+        ofType: SequenceExprSyntax.self,
+        stoppingIf: {
+          $0.kind == .codeBlockItem || $0.kind == .memberBlockItem
+        }
+      )
+    else {
+      return nil
+    }
+
+    return seqExpr
+  }
+}
+
+fileprivate extension BinaryOperatorExprSyntax {
+  var isConcat: Bool {
+    self.operator.tokenKind == .binaryOperator("+")
+  }
+}
+
+fileprivate extension StringLiteralExprSyntax {
+  var isSingleLine: Bool {
+    openingQuote.tokenKind == .stringQuote
+  }
+
+  func isPoundsNotEqual(to otherPounds: TokenSyntax) -> Bool {
+    openingPounds?.tokenKind != otherPounds.tokenKind
+  }
```

Comment:
I think you’re only using this once and the implementation is quite simple as well, so I’d just inline it. I feel like it’s easier to understand the code that way because I don’t need to look up what `isPoundsNotEqual` does.